### ISSUE TITLE
`unnecessary_map_or` suggests code that does not compile

### ISSUE THREAD

**Issue by Kobzol** [rust-lang/rust-clippy#14201]
### Summary

The `unnecessary_map_or` lint can produce a suggestion that causes the resulting code to not compile. It suggests replacing `map_or(false, |v| v == x)` with `== Some(x)`, but the problem is that is `x` isn't `Copy`, then `Some(x)` moves `x`, and it then cannot be used further.

### Lint Name

unnecessary_map_or

### Reproducer

I tried this code:

```rust
fn get_foo() -> Option<String> { None }

fn main() {
    let s = String::from("foo");
    let eq = get_foo().map_or(false, |v| v == s); // Flagged with clippy::unnecessary_map_or
    //let eq = get_foo() == Some(s); // Suggested replacement. Does not compile, because it moves `s`

    println!("{s}");
}
```

### Version

```text
rustc 1.84.1 (e71f9a9a9 2025-01-27)
binary: rustc
commit-hash: e71f9a9a98b0faf423844bf0ba7438f29dc27d58
commit-date: 2025-01-27
host: x86_64-unknown-linux-gnu
release: 1.84.1
LLVM version: 19.1.5
```

**Comment by Kobzol** [comment]
I used `as_ref` to solve this issue in the code where the lint was fired. Not sure if it works in all cases, but if the types match, it probably should.

**Comment by spirali** [comment]
@samueltardieu No problem, do not close your PR. I can still learn on this and then compare our solutions or I will try to find another entry task.
